File: sdk/typescript/src/providers/openai.ts
/**
 * OpenAI provider monkey-patch.
 *
 * Target: `OpenAI.chat.completions.create`
 * Min version: openai >= 4.0
 * Response token path: `response.usage.prompt_tokens`, `response.usage.completion_tokens`
 * Model path: `response.model` (or request kwarg `model`)
 */

import { trace } from "@opentelemetry/api";
import * as attributes from "../attributes.js";
import { getPipelineId, getStage } from "../context.js";
import type { PricingModel } from "../pricing.js";

const MIN_VERSION = "4.0";
const PROVIDER_NAME = "openai";

const tracer = trace.getTracer("hikari");

interface OpenAIUsage {
  prompt_tokens?: number;
  completion_tokens?: number;
}

interface OpenAIResponse {
  usage?: OpenAIUsage;
  model?: string;
}

interface OpenAIClient {
  chat: {
    completions: {
      create: (...args: unknown[]) => Promise<OpenAIResponse>;
    };
  };
}

const originals = new Map<string, unknown>();

function extractTokens(response: OpenAIResponse): [number | null, number | null] {
  const usage = response.usage;
  if (!usage) {
    return [null, null];
  }
  return [usage.prompt_tokens ?? null, usage.completion_tokens ?? null];
}

function extractModel(response: OpenAIResponse, kwargs: Record<string, unknown>): string {
  return response.model ?? (kwargs.model as string) ?? "unknown";
}

function makeWrapper(
  original: (...args: unknown[]) => Promise<OpenAIResponse>,
  pricing: PricingModel
): (...args: unknown[]) => Promise<OpenAIResponse> {
  return async function (this: unknown, ...args: unknown[]): Promise<OpenAIResponse> {
    const spanName = "openai.chat.completions.create";
    const stage = getStage() ?? spanName;

    return tracer.startActiveSpan(spanName, async (span) => {
      try {
        const response = await original.apply(this, args);

        try {
          const [inputTokens, outputTokens] = extractTokens(response);
          const kwargs = (args[0] as Record<string, unknown>) ?? {};
          const model = extractModel(response, kwargs);
          const pipelineId = getPipelineId();

          if (pipelineId) {
            span.setAttribute(attributes.PIPELINE_ID, pipelineId);
          }
          span.setAttribute(attributes.STAGE, stage);
          span.setAttribute(attributes.MODEL, model);
          span.setAttribute(attributes.PROVIDER, PROVIDER_NAME);

          if (inputTokens !== null) {
            span.setAttribute(attributes.TOKENS_INPUT, inputTokens);
          }
          if (outputTokens !== null) {
            span.setAttribute(attributes.TOKENS_OUTPUT, outputTokens);
          }

          const { inputCost, outputCost, totalCost } = pricing.computeCost(
            PROVIDER_NAME,
            model,
            inputTokens,
            outputTokens
          );

          if (inputCost !== null) {
            span.setAttribute(attributes.COST_INPUT, inputCost);
          }
          if (outputCost !== null) {
            span.setAttribute(attributes.COST_OUTPUT, outputCost);
          }
          if (totalCost !== null) {
            span.setAttribute(attributes.COST_TOTAL, totalCost);
          }
        } catch (err) {
          console.debug("Failed to set span attributes:", err);
        }

        span.end();
        return response;
      } catch (err) {
        span.end();
        throw err;
      }
    });
  };
}

/**
 * Apply monkey-patch to OpenAI client methods.
 *
 * @param pricing - The pricing model to use for cost computation
 * @returns `true` if patch was applied, `false` if skipped
 */
export function patch(pricing: PricingModel): boolean {
  try {
    // Dynamic import to avoid hard dependency
    const openai = require("openai");
    const version = openai.VERSION ?? openai.default?.VERSION ?? "0.0.0";

    // Version check
    const major = parseInt(version.split(".")[0] ?? "0", 10);
    if (major < 4) {
      console.warn(`openai version ${version} < ${MIN_VERSION}, skipping`);
      return false;
    }

    // Patch the OpenAI class prototype
    const OpenAI = openai.default ?? openai;
    const proto = OpenAI.prototype;

    if (!proto?.chat?.completions) {
      console.warn("OpenAI client structure not recognized, skipping");
      return false;
    }

    const original = proto.chat.completions.create.bind(proto.chat.completions);
    originals.set("create", original);

    proto.chat.completions.create = makeWrapper(
      original as (...args: unknown[]) => Promise<OpenAIResponse>,
      pricing
    );

    console.info(`Patched openai ${version}`);
    return true;
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "MODULE_NOT_FOUND") {
      console.debug("openai not installed, skipping");
      return false;
    }
    console.warn("Failed to patch openai:", err);
    return false;
  }
}

/**
 * Restore original OpenAI methods.
 */
export function unpatch(): void {
  try {
    const openai = require("openai");
    const OpenAI = openai.default ?? openai;
    const proto = OpenAI.prototype;

    const original = originals.get("create");
    if (original && proto?.chat?.completions) {
      proto.chat.completions.create = original;
      originals.delete("create");
    }
  } catch {
    // Ignore if openai not installed
  }
}

File: sdk/python/src/hikari/providers/openai.py
"""OpenAI provider monkey-patch.

Target: ``openai.resources.chat.completions.Completions.create``
         ``openai.resources.chat.completions.AsyncCompletions.create``
Min version: openai >= 1.0
Response token path: ``response.usage.prompt_tokens``, ``response.usage.completion_tokens``
Model path: ``response.model`` (or request kwarg ``model``)
"""

from __future__ import annotations

import asyncio
import functools
import logging
from typing import Any, Callable

from opentelemetry import trace

from hikari import attributes
from hikari.context import get_pipeline_id, get_stage

logger = logging.getLogger("hikari.providers.openai")

MIN_VERSION = "1.0"
PROVIDER_NAME = "openai"
_originals: dict[str, Any] = {}

tracer = trace.get_tracer("hikari")


def _extract_tokens(response: Any) -> tuple[int | None, int | None]:
    """Extract token counts from an OpenAI response."""
    usage = getattr(response, "usage", None)
    if usage is None:
        return (None, None)
    return (
        getattr(usage, "prompt_tokens", None),
        getattr(usage, "completion_tokens", None),
    )


def _extract_model(response: Any, kwargs: dict[str, Any]) -> str:
    """Extract the model identifier from response or request kwargs."""
    model = getattr(response, "model", None)
    if model is None:
        model = kwargs.get("model", "unknown")
    return model


def _make_sync_wrapper(
    original: Callable[..., Any],
    pricing_model: Any,
) -> Callable[..., Any]:
    @functools.wraps(original)
    def wrapper(self: Any, *args: Any, **kwargs: Any) -> Any:
        span_name = "openai.chat.completions.create"
        stage = get_stage() or span_name
        with tracer.start_as_current_span(span_name) as span:
            try:
                response = original(self, *args, **kwargs)
            except Exception:
                raise
            else:
                try:
                    input_tokens, output_tokens = _extract_tokens(response)
                    model = _extract_model(response, kwargs)
                    pipeline_id = get_pipeline_id()
                    if pipeline_id:
                        span.set_attribute(attributes.PIPELINE_ID, pipeline_id)
                    span.set_attribute(attributes.STAGE, stage)
                    span.set_attribute(attributes.MODEL, model)
                    span.set_attribute(attributes.PROVIDER, PROVIDER_NAME)
                    if input_tokens is not None:
                        span.set_attribute(attributes.TOKENS_INPUT, input_tokens)
                    if output_tokens is not None:
                        span.set_attribute(attributes.TOKENS_OUTPUT, output_tokens)

                    input_cost, output_cost, total_cost = pricing_model.compute_cost(
                        PROVIDER_NAME, model, input_tokens, output_tokens
                    )
                    if input_cost is not None:
                        span.set_attribute(attributes.COST_INPUT, input_cost)
                    if output_cost is not None:
                        span.set_attribute(attributes.COST_OUTPUT, output_cost)
                    if total_cost is not None:
                        span.set_attribute(attributes.COST_TOTAL, total_cost)
                except Exception:
                    logger.debug("Failed to set span attributes", exc_info=True)
                return response

    return wrapper


def _make_async_wrapper(
    original: Callable[..., Any],
    pricing_model: Any,
) -> Callable[..., Any]:
    @functools.wraps(original)
    async def wrapper(self: Any, *args: Any, **kwargs: Any) -> Any:
        span_name = "openai.chat.completions.create"
        stage = get_stage() or span_name
        with tracer.start_as_current_span(span_name) as span:
            try:
                response = await original(self, *args, **kwargs)
            except Exception:
                raise
            else:
                try:
                    input_tokens, output_tokens = _extract_tokens(response)
                    model = _extract_model(response, kwargs)
                    pipeline_id = get_pipeline_id()
                    if pipeline_id:
                        span.set_attribute(attributes.PIPELINE_ID, pipeline_id)
                    span.set_attribute(attributes.STAGE, stage)
                    span.set_attribute(attributes.MODEL, model)
                    span.set_attribute(attributes.PROVIDER, PROVIDER_NAME)
                    if input_tokens is not None:
                        span.set_attribute(attributes.TOKENS_INPUT, input_tokens)
                    if output_tokens is not None:
                        span.set_attribute(attributes.TOKENS_OUTPUT, output_tokens)

                    input_cost, output_cost, total_cost = pricing_model.compute_cost(
                        PROVIDER_NAME, model, input_tokens, output_tokens
                    )
                    if input_cost is not None:
                        span.set_attribute(attributes.COST_INPUT, input_cost)
                    if output_cost is not None:
                        span.set_attribute(attributes.COST_OUTPUT, output_cost)
                    if total_cost is not None:
                        span.set_attribute(attributes.COST_TOTAL, total_cost)
                except Exception:
                    logger.debug("Failed to set span attributes", exc_info=True)
                return response

    return wrapper


def patch(pricing_model: Any) -> bool:
    """Apply monkey-patch to OpenAI client methods.

    Returns ``True`` if patch was applied, ``False`` if skipped.
    """
    try:
        import openai  # noqa: F811
    except ImportError:
        logger.debug("openai not installed, skipping")
        return False

    try:
        from packaging.version import Version
    except ImportError:
        # Fallback: try parsing version manually
        version_str = getattr(openai, "__version__", "0.0.0")
        major = int(version_str.split(".")[0])
        if major < 1:
            logger.warning("openai version %s < %s, skipping", version_str, MIN_VERSION)
            return False
    else:
        version_str = getattr(openai, "__version__", "0.0.0")
        if Version(version_str) < Version(MIN_VERSION):
            logger.warning("openai version %s < %s, skipping", version_str, MIN_VERSION)
            return False

    try:
        from openai.resources.chat.completions import AsyncCompletions, Completions

        _originals["sync_create"] = Completions.create
        _originals["async_create"] = AsyncCompletions.create

        Completions.create = _make_sync_wrapper(Completions.create, pricing_model)  # type: ignore[assignment]
        AsyncCompletions.create = _make_async_wrapper(AsyncCompletions.create, pricing_model)  # type: ignore[assignment]
        logger.info("Patched openai %s", version_str)
        return True
    except Exception:
        logger.warning("Failed to patch openai", exc_info=True)
        return False


def unpatch() -> None:
    """Restore original OpenAI methods."""
    try:
        from openai.resources.chat.completions import AsyncCompletions, Completions

        if "sync_create" in _originals:
            Completions.create = _originals.pop("sync_create")  # type: ignore[assignment]
        if "async_create" in _originals:
            AsyncCompletions.create = _originals.pop("async_create")  # type: ignore[assignment]
    except ImportError:
        pass
